cmd/snap-bootstrap/cmd_initramfs_mounts_cvm.go

```diff
@@ -44,9 +49,175 @@ func (*genericCVMModel) Grade() asserts.ModelGrade {
+	foundReadOnlyPartition := ""
+	foundWritablePartition := ""
+
+	partitionMounts := []partitionMount{}
+
+	// Configure the overlay filesystems mounts from the manifest.
+	for _, p := range im.Partitions {
+		pm := partitionMount{
+			Opts: &systemdMountOptions{},
+		}
+
+		pm.GptLabel = p.GptLabel
+
+		// All detected partitions are mounted by default under /run/mnt/<GptLabel of partition>
+		pm.Where = filepath.Join(boot.InitramfsRunMntDir, p.GptLabel)
+
+		if p.Overlay == "lowerdir" {
+			// XXX: currently only one lower layer is permitted. The rest, if found, are ignored.
+			if foundReadOnlyPartition != "" {
+				continue
+			}
+			foundReadOnlyPartition = pm.GptLabel
+
+			// systemd-mount will run fsck by default when attempting to mount the partition and potentially corrupt it.
+			// This will cause dm-verity to fail when attempting to set up the dm-verity mount.
+			// fsck should be/is run by the encrypt-cloud-image tool prior to generating dm-verity data.
+			pm.Opts.NeedsFsck = false
+			pm.Opts.VerityRootHash = p.RootHash
+
+			// Auto-discover verity device from disk for partition types meant to be used as lowerdir
+			verityPartition, err := disk.FindMatchingPartitionWithPartLabel(p.GptLabel + "-verity")
+			if err != nil {
+				return []partitionMount{}, err
+			}
+			pm.Opts.VerityHashDevice = verityPartition.KernelDeviceNode
+		} else {
+			// Only one writable partition is permitted.
+			if foundWritablePartition != "" {
+				continue
+			}
+			// Manifest contains a partition meant to be used as a writable overlay for the non-ephemeral vm case.
+			// If it is encrypted, its key will be autodiscovered based on its FsLabel later.
+			foundWritablePartition = p.GptLabel
+			pm.Opts.NeedsFsck = true
+		}
+
+		partitionMounts = append(partitionMounts, pm)
+	}
+
+	if foundReadOnlyPartition == "" {
+		return nil, fmt.Errorf("manifest doesn't contain any partition with Overlay type lowerdir")
+	}
+
+	// If no writable partitions were found in the manifest, Configure a tmpfs filesystem for the upper and workdir layers
+	// of the final rootfs mount.
+	if foundWritablePartition == "" {
+		foundWritablePartition = "writable-tmp"
+
+		pm := partitionMount{
+			Where:    filepath.Join(boot.InitramfsRunMntDir, "writable-tmp"),
+			GptLabel: "writable-tmp",
+			Opts: &systemdMountOptions{
+				Tmpfs: true,
+			},
+		}
+
+		partitionMounts = append(partitionMounts, pm)
+	}
+
+	// Configure the merged overlay filesystem mount.
+	pm := partitionMount{
+		Where:    boot.InitramfsDataDir,
+		GptLabel: "cloudimg-rootfs",
+		Opts: &systemdMountOptions{
+			Overlayfs: true,
+			LowerDirs: []string{filepath.Join(boot.InitramfsRunMntDir, partitionMounts[0].GptLabel)},
+		},
+	}
+
+	pm.Opts.UpperDir = filepath.Join(boot.InitramfsRunMntDir, foundWritablePartition, "upper")
+	pm.Opts.WorkDir = filepath.Join(boot.InitramfsRunMntDir, foundWritablePartition, "work")
```

Comment:
This could be done when `pm` is defined, splitting the initialization of Opts does not seem necessary.

Reply:
Fixed.